tracer/src/Datadog.Trace/Debugger/DebuggerManager.cs

```diff
@@ -0,0 +1,343 @@
+// <copyright file="DebuggerManager.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Datadog.Trace.Agent.DiscoveryService;
+using Datadog.Trace.ClrProfiler;
+using Datadog.Trace.Debugger.ExceptionAutoInstrumentation;
+using Datadog.Trace.Debugger.Sink;
+using Datadog.Trace.Debugger.Snapshots;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Processors;
+using Datadog.Trace.RemoteConfigurationManagement;
+using Datadog.Trace.Telemetry;
+using Datadog.Trace.Telemetry.Metrics;
+
+#nullable enable
+
+namespace Datadog.Trace.Debugger
+{
+    internal class DebuggerManager
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(DebuggerManager));
+        private static readonly Lazy<DebuggerManager> _lazyInstance =
+            new(
+                () => new DebuggerManager(DebuggerSettings.FromDefaultSource(), ExceptionReplaySettings.FromDefaultSource()),
+                LazyThreadSafetyMode.ExecutionAndPublication);
+
+        private readonly SemaphoreSlim _semaphore;
+        private readonly CancellationTokenSource _cancellationToken;
+        private volatile bool _isShuttingDown;
+        private int _initialized;
+
+        private DebuggerManager(DebuggerSettings debuggerSettings, ExceptionReplaySettings exceptionReplaySettings)
+        {
+            _initialized = 0;
+            _isShuttingDown = false;
+            _semaphore = new SemaphoreSlim(1, 1);
```

Comment:
You bound the semaphore to max of `1`. It results in rejections of subsequent threads. We should treat the latest configuration update as the most relevant one and avoid rejecting it.

Reply:
Fixed in 04e5faa8e2c7d651fa234877c1b0c29a27689ac3